### PR TITLE
Rhinofi

### PR DESCRIPTION
**Pull request by EvgeniiaVak**

### CODE REVIEW

File: protocols/rhinofi.json [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	],
+	"logo_dark": "/protocols/rhinofi-logo-dark.svg",
+	"logo_light": "/protocols/rhinofi-logo-light.svg",
+	"icon_url": "/protocols/rhinofi-icon-dark.png",
+	"icon_dark": "/protocols/rhinofi-icon-dark.png",
+	"icon_light": "/protocols/rhinofi-icon-light.png",
+	"description": "rhino.fi is a Layer 2 bridge for effortless multi-chain token movement",
+	"mainnet": "yes",
+	"type": "bridge",
+	"chains_supported": "Ethereum, Arbitrum, BSC, Polygon, zkSync, zkEVM, Optimism, Linea, Base, Manta, Starknet, Scroll",
```
Comment (konradstrachan) [inline protocols/rhinofi.json:41]: I think a few are missing here?

File: protocols/rhinofi.json [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	"xchain_sync": "Canonical bridges for message passing",
+	"auction_design": "-",
+	"auction_design_ranking": "Poor",
+	"auction_design_details": "The bridging process involves the user giving rhino.fi assets on their origin chain and rhino.fi providing their desired assets on their destination chain. There is always only one filler - rhino.fi.",
+	"auction_openness": "Poor",
+	"auction_openness_details": "There is always only one filler - rhino.fi.",
```
Comment (konradstrachan) [inline protocols/rhinofi.json:47]: I'd rephrase as : The protocol operates a single filler that executes all orders

File: protocols/rhinofi.json [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	"filler_permissionlessness": "Poor",
+	"filler_permissionlessness_details": "There is always only one filler - rhino.fi. Liquidity is provided by rhino.fi AMMs: each liquidity pool includes a pair of tokens, as well as a third, liquidity provider (LP) token. The pool is filled by the community, in exchange for a share of trading fees.",
```
Comment (konradstrachan) [inline protocols/rhinofi.json:49]: This is not correct.

I'd write it as: No mechanism exists for other fillers to solve orderflow originating from Rhino.fi as the order expression information is incomplete and no mechanism exists for refunding of inventory filled by other entities.

File: protocols/rhinofi.json [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	"censorship_resistance": "Poor",
+	"censorship_resistance_details": "Because rhino.fi is the only filler, there is a possibility for censorship.",
```
Comment (konradstrachan) [inline protocols/rhinofi.json:51]: I'd rephrase as: Since the only running filler belongs to the protocol, there is no censorship resistance.

File: protocols/rhinofi.json [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	"filler_failure_resistance": "Good",
+	"filler_failure_resistance_details": "Orders are filled by smart contracts from the pre-populated liquidity pools on the destination chain. If there is not enough liquidity on the destination chain, the order cannot be accepted. If the order is accepted it will be filled."
```
Comment (konradstrachan) [inline protocols/rhinofi.json:53]: This is wrong.

"filler_failure_resistance": Average

I'd rephrase as: Since the only running filler belongs to the protocol, were this or the off-chain infra to fail there would be no way for a user to complete a bridging operation. Rhino.fi is built on StarkEx which provides some mechanism for users to withdraw funds to Ethereum L1 in the event the operator goes offline providing some limited recovery options.

File: protocol_data/rhinofi.py [hunk 6]
```diff
@@ -0,0 +1,687 @@
...
+    # fill event filter will not result in reverted txs being picked up
+    # TODO: verify rhino.fi does not have any not-successful fills, because
+    # there is only one filler and it gets the funds from pre-filled pools, related discord chat:
+    # https://discord.com/channels/745570257808130058/745571091438764062/1265924362276438108
+    return { 0 : [] }
+
+def get_fill_event_filter():
+    # To record fill events specify the event name
+    # NOTE: wherever possible please also include the fill function filter above
+    return { 0 : ['BridgedWithdrawal', 'BridgedWithdrawalWithData', 'BridgedWithdrawalWithNative'] }
```
Comment (konradstrachan) [inline protocol_data/rhinofi.py:687]: Not seeing BridgedWithdrawalWithData in the official repo, did you see this onchain?

**Review by konradstrachan — CHANGES_REQUESTED** [review]
Changes requested

File: protocols/rhinofi.json (hunk 2 above)
Comment (EvgeniiaVak) [reply protocols/rhinofi.json]: fixed

File: protocols/rhinofi.json (hunk 3 above)
Comment (EvgeniiaVak) [reply protocols/rhinofi.json]: fixed

File: protocols/rhinofi.json (hunk 4 above)
Comment (EvgeniiaVak) [reply protocols/rhinofi.json]: fixed

File: protocols/rhinofi.json (hunk 5 above)
Comment (EvgeniiaVak) [reply protocols/rhinofi.json]: fixed

File: protocol_data/rhinofi.py (hunk 6 above)
Comment (EvgeniiaVak) [reply protocol_data/rhinofi.py:687]: I took it from the ABI (I took the ABI from contract on binance scanner). Maybe I used incorrect ABI? Let me know if there is a better way to find the ABI. Here is how I found it:
Screenshot 2024-08-01 at 14 59 09
Screenshot 2024-08-01 at 14 59 32
Screenshot 2024-08-01 at 15 00 10

File: protocols/rhinofi.json (hunk 1 above)
Comment (konradstrachan) [reply protocols/rhinofi.json:41]: Ethereum is a good point because the way deposits work from there is quite different as it links in to the StarkEx contract

File: protocol_data/rhinofi.py (hunk 6 above)
Comment (konradstrachan) [reply protocol_data/rhinofi.py:687]: Looking at https://bscscan.com/address/0x3337b457dad8339a90a7ba4fdd1d859915b96f7f#code I don't see this event in the contract?

Comment (konradstrachan) [reply protocol_data/rhinofi.py:687]: Although this makes an interesting point.. perhaps this is an earlier version of the contract which is not longer in use, but we will need the events to backfill the data

Comment (konradstrachan) [reply protocol_data/rhinofi.py:687]: Found it at https://bscscan.com/address/0x937c802934c71b716879c52c91c2884a36b448fb

Comment (EvgeniiaVak) [reply protocol_data/rhinofi.py:687]: ok, so we are keeping them here, right?

File: protocol_data/rhinofi.py [hunk 7]
```diff
@@ -0,0 +1,687 @@
...
+        },
+        '534352': {
+            'deposit': { 694 : '0x87627c7E586441EeF9eE3C28B66662e897513f33' },
+            'fill': { 694 : '0x87627c7E586441EeF9eE3C28B66662e897513f33' }
+        },
+    }
+    return contracts[chain_id][type]
+
+# TODO: find solana contracts, they are absent in https://github.com/rhinofi/contracts_public/
+def get_function_identifier(chain_id, type):
```
Comment (konradstrachan) [inline protocol_data/rhinofi.py:60]: I've asked about this, let's hold off until we get a response

Comment (EvgeniiaVak) [reply protocol_data/rhinofi.py:60]: solana won't work anyways - we match based on address, solana addresses are not evm compatible
